subprojects/core/src/main/java/org/gradle/internal/buildevents/BuildExceptionReporter.java

```diff
@@ -164,6 +167,18 @@ private void renderMultipleBuildExceptions(Failure failure) {
         }
     }
 
+    private static String extractGranularity(Failure failure) {
+        int depth = 0;
+        Throwable cause = failure.getOriginal();
+        while (cause != null && depth++ < MAX_GRANULARITY_SEARCH_DEPTH) {
```

Comment:
💭 This logic is a bit weird, do we have any better way to extract granularity here?